File: apps/flutter/lib/ui/copy.dart
/// User-facing copy, mirrored from `/shared/copy/errors.json`, keyed by the taxonomy code.
library;

import '../core/core.dart';

class ErrorCopy {
  final String title;
  final String message;
  final String? actionLabel;
  const ErrorCopy(this.title, this.message, this.actionLabel);
}

const _byCode = <String, (String, String)>{
  'NetworkUnavailable': ('No connection', 'You appear to be offline. Check your connection and try again.'),
  'Timeout': ('Taking too long', 'The server did not answer in time. It may still be working.'),
  'Unauthorized': ('Signed out', 'Your session has ended. Sign in again to continue.'),
  'Forbidden': ('Not allowed', 'You do not have access to this.'),
  'RateLimited': ('Slow down', 'Too many requests. Wait a moment before trying again.'),
  'Conflict': ('No longer available', 'That is gone or has changed. Refresh and pick again.'),
  'Validation': ('Check your details', 'Some of what you entered was not accepted.'),
  'ServerError': ('Something broke', 'The server had a problem. This is not your fault.'),
  'MalformedResponse': ('Unexpected response', 'We received something we could not read and stopped to be safe.'),
  'PaymentDeclined': ('Payment declined', 'Your payment was declined. Try a different method.'),
  'PaymentUnknown': ('Confirming payment', 'We are checking with the payment provider. Do not pay again.'),
  'Unknown': ('Something went wrong', 'An unexpected error occurred.'),
};

ErrorCopy copyFor(AppError error) {
  final entry = _byCode[error.code] ?? _byCode['Unknown']!;
  return ErrorCopy(entry.$1, entry.$2, _actionLabel(error.recovery));
}

String? _actionLabel(Recovery recovery) => switch (recovery) {
      Recovery.retry => 'Try again',
      Recovery.back => 'Go back',
      Recovery.refresh => 'Refresh',
      Recovery.signIn => 'Sign in',
      Recovery.wait => 'Keep waiting',
      Recovery.none => null,
    };

File: apps/react-native/src/ui/copy.ts
// User-facing copy, mirrored from `/shared/copy/errors.json`, keyed by the taxonomy code.

import { AppError, ErrorCode, Recovery } from '../core/core';

export interface ErrorCopy {
  title: string;
  message: string;
  actionLabel: string | null;
}

const BY_CODE: Record<ErrorCode, { title: string; message: string }> = {
  NetworkUnavailable: { title: 'No connection', message: 'You appear to be offline. Check your connection and try again.' },
  Timeout: { title: 'Taking too long', message: 'The server did not answer in time. It may still be working.' },
  Unauthorized: { title: 'Signed out', message: 'Your session has ended. Sign in again to continue.' },
  Forbidden: { title: 'Not allowed', message: 'You do not have access to this.' },
  RateLimited: { title: 'Slow down', message: 'Too many requests. Wait a moment before trying again.' },
  Conflict: { title: 'No longer available', message: 'That is gone or has changed. Refresh and pick again.' },
  Validation: { title: 'Check your details', message: 'Some of what you entered was not accepted.' },
  ServerError: { title: 'Something broke', message: 'The server had a problem. This is not your fault.' },
  MalformedResponse: { title: 'Unexpected response', message: 'We received something we could not read and stopped to be safe.' },
  PaymentDeclined: { title: 'Payment declined', message: 'Your payment was declined. Try a different method.' },
  PaymentUnknown: { title: 'Confirming payment', message: 'We are checking with the payment provider. Do not pay again.' },
  Unknown: { title: 'Something went wrong', message: 'An unexpected error occurred.' },
};

const ACTION_LABEL: Record<Recovery, string | null> = {
  retry: 'Try again',
  back: 'Go back',
  refresh: 'Refresh',
  signIn: 'Sign in',
  wait: 'Keep waiting',
  none: null,
};

export function copyFor(error: AppError): ErrorCopy {
  const { title, message } = BY_CODE[error.code] ?? BY_CODE.Unknown;
  return { title, message, actionLabel: ACTION_LABEL[error.recovery] };
}
